drivers/pinctrl/pinctrl_stm32.c

```diff
@@ -0,0 +1,574 @@
+	if (pin >= pinctrl_tmpl_control_get_pins_count(dev)) {
+		return -ENOTSUP;
+	}
+
+	*config = 0;
+
+	GPIO_TypeDef *ll_gpio_port = pinctrl_stm32_ll_gpio_port(pin);
+	u32_t ll_gpio_pin = pinctrl_stm32_ll_gpio_pin(pin);
+
+	/* current pin mode (muxing function) */
+	switch (LL_GPIO_GetPinMode(ll_gpio_port, ll_gpio_pin)) {
+	case LL_GPIO_MODE_ALTERNATE:
+		*config |= PINCTRL_CONFIG_OUTPUT_ENABLE |
+			   PINCTRL_CONFIG_INPUT_ENABLE |
+			   PINCTRL_CONFIG_INPUT_SCHMITT_ENABLE;
+		break;
+	case LL_GPIO_MODE_OUTPUT:
+		*config |= PINCTRL_CONFIG_OUTPUT_ENABLE |
+			   PINCTRL_CONFIG_INPUT_ENABLE |
+			   PINCTRL_CONFIG_INPUT_SCHMITT_ENABLE;
+		break;
+	case LL_GPIO_MODE_INPUT:
+		*config |= PINCTRL_CONFIG_INPUT_ENABLE |
+			   PINCTRL_CONFIG_OUTPUT_DISABLE |
+			   PINCTRL_CONFIG_INPUT_SCHMITT_ENABLE;
+		break;
+	case LL_GPIO_MODE_ANALOG:
+		*config |= PINCTRL_CONFIG_OUTPUT_DISABLE |
+			   PINCTRL_CONFIG_INPUT_DISABLE |
+			   PINCTRL_CONFIG_INPUT_SCHMITT_DISABLE |
+			   PINCTRL_CONFIG_BIAS_DISABLE;
+		break;
+	}
+
+	if (!(*config | PINCTRL_CONFIG_BIAS_DISABLE)) {
+		switch (LL_GPIO_GetPinOutputType(ll_gpio_port, ll_gpio_pin)) {
+		case LL_GPIO_OUTPUT_PUSHPULL:
+			*config |= PINCTRL_CONFIG_BIAS_DISABLE;
+			break;
+		case LL_GPIO_PULL_UP:
+			*config |= PINCTRL_CONFIG_BIAS_PULL_UP;
+			break;
+		case LL_GPIO_PULL_DOWN:
+			*config |= PINCTRL_CONFIG_BIAS_PULL_DOWN;
+			break;
+		}
+	}
+
+	switch (LL_GPIO_GetPinSpeed(ll_gpio_port, ll_gpio_pin)) {
+	case LL_GPIO_SPEED_FREQ_LOW:
+		*config |= PINCTRL_CONFIG_SPEED_SLOW;
+		break;
+	case LL_GPIO_SPEED_FREQ_MEDIUM:
+		*config |= PINCTRL_CONFIG_SPEED_MEDIUM;
+		break;
+	case LL_GPIO_SPEED_FREQ_HIGH:
+		*config |= PINCTRL_CONFIG_SPEED_HIGH;
+		break;
+	}
+
+	switch (LL_GPIO_GetPinOutputType(ll_gpio_port, ll_gpio_pin)) {
+	case LL_GPIO_OUTPUT_PUSHPULL:
+		*config |= PINCTRL_CONFIG_DRIVE_PUSH_PULL;
+		break;
+	case LL_GPIO_OUTPUT_OPENDRAIN:
+		*config |= PINCTRL_CONFIG_DRIVE_OPEN_DRAIN;
+		break;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Configure a pin.
+ *
+ * The following configuration options are supported:
+ *   - @ref PINCTRL_CONFIG_BIAS_DISABLE
+ *   - @ref PINCTRL_CONFIG_BIAS_PULL_UP
+ *   - @ref PINCTRL_CONFIG_BIAS_PULL_DOWN
+ *   - @ref PINCTRL_CONFIG_BIAS_PULL_PIN_DEFAULT
+ *   - @ref PINCTRL_CONFIG_DRIVE_PUSH_PULL
+ *   - @ref PINCTRL_CONFIG_DRIVE_OPEN_DRAIN
+ *   - @ref PINCTRL_CONFIG_SPEED_SLOW
+ *   - @ref PINCTRL_CONFIG_SPEED_MEDIUM
+ *   - @ref PINCTRL_CONFIG_SPEED_FAST
+ *   - @ref PINCTRL_CONFIG_SPEED_HIGH
+ *
+ * @note For safety purposes the pin should be reset to input before
+ * configuration.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param pin Pin
+ * @param config
+ * @retval 0 on success
+ * @retval -ENOTSUP if requested pin is not available on this controller
+ */
+static int pinctrl_stm32_config_set(struct device *dev, u16_t pin, u32_t config)
+{
+	ARG_UNUSED(dev);
+
+	if (pin >= pinctrl_tmpl_control_get_pins_count(dev)) {
+		return -ENOTSUP;
+	}
```

Comment:
I believe that it is by-design a convention to use __ASSERT for input validation like this.

The idea is that it should be up to the user whether he wants safety or speed.

__ASSERT(pin < pinctrl_tmpl_control_get_pins_count(dev))

would enable this.

Reply:
> safety or speed

I´m not shure here. At least by config_set you can burn the HW. Speed is rarely an issue for pin configuration.